code/messaging/src/main/java/com/adobe/marketing/mobile/messaging/internal/EdgePersonalizationResponseHandler.java

```diff
@@ -163,128 +199,300 @@ void fetchMessages(final List<Surface> surfaces) {
+
+        // handle in-app message rules
+        if (!MessagingUtils.isNullOrEmpty(inAppRules)) {
+            Log.trace(MessagingConstants.LOG_TAG, SELF_TAG, "The personalization:decisions response contains InApp message definitions.");
+        }
+        if (clearExistingRules) {
+            Log.debug(LOG_TAG, SELF_TAG, "Loading %d rule(s) into the rules engine.", inAppRules != null ? inAppRules.size() : 0);
+            launchRulesEngine.replaceRules(inAppRules);
+        } else {
+            Log.debug(LOG_TAG, SELF_TAG, "Added %d rule(s) into the rules engine.", inAppRules != null ? inAppRules.size() : 0);
+            launchRulesEngine.addRules(inAppRules);
         }
 
-        final String appSurface = getAppSurface();
-        Log.trace(LOG_TAG, SELF_TAG, "Loading in-app message definitions from personalization:decisions network response.");
-        processPropositions(propositions, clearExistingRules, true, appSurface);
+        // handle feed rules
+        if (!MessagingUtils.isNullOrEmpty(feedRules)) {
+            Log.trace(MessagingConstants.LOG_TAG, SELF_TAG, "The personalization:decisions response contains feed message definitions.");
+        }
+        if (clearExistingRules) {
+            Log.debug(LOG_TAG, SELF_TAG, "Loading %d rule(s) into the feed rules engine.", feedRules != null ? feedRules.size() : 0);
+            feedRulesEngine.replaceRules(feedRules);
+        } else {
+            Log.debug(LOG_TAG, SELF_TAG, "Added %d rule(s) into the feed rules engine.", feedRules != null ? feedRules.size() : 0);
+            feedRulesEngine.addRules(feedRules);
+        }
+
+        inboundMessages = feedRulesEngine.evaluate(edgeResponseEvent);
+        if (!MapUtils.isNullOrEmpty(inboundMessages)) {
+            updateInboundMessages(inboundMessages, requestedSurfaces);
+        }
+
+        final Map<Surface, List<Proposition>> requestedPropositions = retrievePropositions(requestedSurfaces);
+        if (MapUtils.isNullOrEmpty(requestedPropositions)) {
+            Log.trace(MessagingConstants.LOG_TAG, SELF_TAG, "Not dispatching a notification event, personalization:decisions response does not contain propositions.");
+            return;
+        }
+
+        // dispatch an event with the propositions received from the remote
+        final Map<String, Object> eventData = new HashMap<>();
+        final Map<String, Object> requestedPropositionsMap = new HashMap<>();
+        for (final Map.Entry<Surface, List<Proposition>> propositionEntry : requestedPropositions.entrySet()) {
+            final List<Map<String, Object>> convertedPropositions = new ArrayList<>();
+            for (final Proposition proposition : propositionEntry.getValue()) {
+                convertedPropositions.add(proposition.toEventData());
+            }
+            requestedPropositionsMap.put(propositionEntry.getKey().getUri(), convertedPropositions);
+        }
+        eventData.put(PROPOSITIONS, requestedPropositionsMap);
```

Comment:
Let us keep the events format consistent across platforms. For propositions in event data, we are passing the list of propositions here in iOS.